cmd/thanos/main.go

```diff
@@ -288,10 +298,58 @@ func defaultGRPCServerOpts(logger log.Logger, cert, key, clientCA string) ([]grp
 		tlsCfg.ClientCAs = certPool
 		tlsCfg.ClientAuth = tls.RequireAndVerifyClientCert
 
-		level.Info(logger).Log("msg", "gRPC server TLS client verification enabled")
+		level.Info(logger).Log("msg", "server TLS client verification enabled")
+	}
+
+	return tlsCfg, nil
+}
+
+func defaultTLSClientOpts(logger log.Logger, secure bool, cert, key, caCert, serverName string) (*tls.Config, error) {
+	if !secure {
```

Comment:
I know this does not make sense from a high level, but I think... let's remove this. With this change if someone use `secure false ` and key and cert set we silently ignore key and cert. I think failing would be better.

Reply:
in that case should we remove the `-secure` flag as well? I also think it is confusing to explicitly have to opt into client TLS in addition to specifying the certs. but it's a breaking change

Reply:
I think yes - `secure` only makes sense if you have certs (?)

Reply:
hmm I'm looking at this again and I think the actual problem here is that every gRPC ClientConn created by the querier re-uses the same set of `[]DialOpt`, which must be either secure OR insecure, but not both. For this reason, the `grpc-client-tls-secure` flag is used to toggle off ALL client TLS features. Another option would be to use the insecure DialOpt unless ANY client TLS flag is provided.